### ISSUE TITLE
Feedback on migration to trait-based design

### ISSUE THREAD

**Issue by madsmtm** [rust-windowing/winit#3626]
Thank you for your interest in expressing your feedback on Winit `v0.30`!

### Background

Winit is moving towards a trait-based API for several reasons, see https://github.com/rust-windowing/winit/issues/3432 for details. This has quite a large impact on how you structure your code, and while the design is not yet optimal, and will need additional work, we feel confident that this is the right way forwards, so in `v0.30`, we've tried to provide a somewhat gentle update path.

### Feedback

We're aware that the API is more verbose, that is to be expected when implementing an entire trait instead of letting the compiler figure out the type annotations of a closure, but we're interested in knowing what kinds of code is made harder or impossible with this new design? Were you using some pattern before that you can no longer do, or no longer easily do? Please state so in a comment on this issue!

---

In this release, we also changed how windows are created, but that is not strictly part of this issue - please [submit a new issue](https://github.com/rust-windowing/winit/issues/new/choose) if you're having trouble with that. Note that we're well aware that `Option<Window>` is a bit cumbersome, in the future, managing the state for windows and creating/destroying them at the right times will likely become easier to do, see https://github.com/rust-windowing/winit/issues/2903 for some of the progress on that.

**Comment by XavierCS-dev** [comment]
I used a mut closure as the users main loop for my game engine, but with this new setup, I am struggling to figure out how to set it up properly without also killing performance, any pointers? I only updated due to the issues with exclusive  full screen on 0.29

**Comment by kchibisov** [comment]
Please look into CHANGELOG and migration guide in it. Closure is just a `struct` where each field is a capture variable. It's also not _enforced_.

**Comment by XavierCS-dev** [comment]
> Please look into CHANGELOG and migration guide in it. Closure is just a `struct` where each field is a capture variable. It's also not _enforced_.

Thanks, I managed to figure it out

**Comment by XavierCS-dev** [comment]
I am also finding I will need to restructure very significant parts of the engine if I am to use the new way of initialising windows. Currently, I use an app struct which contains the main engine state struct, however to initialise wgpu, the window is needed, however you can't get the window until during the event loop so wgpu can't be initialised until then, which means the user can't directly get access to the app struct, and can only get a mutable reference to it. I will also need to write and implement ApplicationHandler<T> individually for each rendering backend.

**Comment by kchibisov** [comment]
But we just swapped `Window::new(event_loop)` to `event_loop.create_new_window()`, so it's the same as well. The only difference is that you need loop running to have _non deprecated_ code, but you can not really work on android, etc if you don't do so anyway, so sounds like a _benefit_ for a cross platform GPU stuff.

**Comment by XavierCS-dev** [comment]
> But we just swapped `Window::new(event_loop)` to `event_loop.create_new_window()`, so it's the same as well. The only difference is that you need loop running to have _non deprecated_ code, but you can not really work on android, etc if you don't do so anyway, so sounds like a _benefit_ for a cross platform GPU stuff.

This is true, I would also rather not use deprecated code. I managed to get it working, very messy but I can fix it later. I am still finding that Exclusive full screen still doesn't work on Hyprland, although I could just be doing it wrong, and besides that is out of scope for this thread I believe.

**Comment by kchibisov** [comment]
Docs clearly say that exclusive fullscreen doens't exist on Wayland.

**Comment by XavierCS-dev** [comment]
> Docs clearly say that exclusive fullscreen doens't exist on Wayland.

It seems I didn't understand the difference between borderless and Exclusive. The borderless mode seems to allow window sizes and resolutions different from the display's resolution, not what I expected from a borderless mode but I guess it is possible.

**Comment by kchibisov** [comment]
exclusive means that you get exclusive control. As in you can also modeset, as in change monitor resolution. Borderless just ordinary fullscreen one.

**Comment by tgross35** [comment]
I do think this is a much cleaner interface, but I really struggled getting something up and running with `wgpu`, because the wgpu `Surface` typically borrows the `Window`, which is easy with the event loop but turns out to be self-referential when you need to construct the `window` within the `resumed` function. This was kind of a pain to get things working, but I wound up with something like this:

```rust
struct App {
    window: Option<Arc<Window>>,
    gfx_state: Option<GfxState>,
}

impl ApplicationHandler<MyUserEvent> for State {
    fn resumed(&mut self, event_loop: &ActiveEventLoop) {
        let window = Arc::new(
            event_loop.create_window(Window::default_attributes() /* ... */).unwrap(),
        );
        self.window = Some(window);
        let gfx_state = GfxState::new(Arc::clone(self.window.as_ref().unwrap()));
        self.gfx_state = Some(gfx_state);
    }

    fn window_event(&mut self, event_loop: &ActiveEventLoop, window_id: WindowId, event: WindowEvent) {
        // ...
        self.gfx_state.as_mut().unwrap().paint() // paint is 
    }
}

struct GfxState {
    device: wgpu::Device,
    surface_desc: wgpu::SurfaceConfiguration,
    surface: wgpu::Surface<'static>, // important part: 'static
    // ... all the other queues / buffers / etc needed
}

impl GfxState {
    fn new(window: Arc<Window>, app_state: &mut AppState) -> Self {
        let instance = wgpu::Instance::default();
        let size = window.inner_size();

        // important part: we can create a `wgpu::Surface<'static>` from an `Arc`
        let surface = instance.create_surface(window).unwrap();

        // ... configure pipelines
    }

    fn paint(&mut self, state: &mut AppState) {
        // perform the rendering
    }
}
```

This could be slightly simplified by putting `Arc<Window>` and `GfxState` in a child struct within the same option. Does anyone have a better pattern?

**Comment by ifsheldon** [comment]
Based on @tgross35 comment, another issue comes up, which is async. Working with wgpu, we have to deal with some `Future`s like below
```rust
let instance = wgpu::Instance::default();
let surface = instance.create_surface(window).expect("Failed to create surface");
// need adapter to create the device and queue
let adapter = instance
    .request_adapter(&wgpu::RequestAdapterOptions {
        power_preference: wgpu::PowerPreference::default(),
        force_fallback_adapter: false,
        compatible_surface: Some(&surface),
    })
    .await
    .unwrap();
let (device, queue) = adapter
    .request_device(
        &wgpu::DeviceDescriptor {
            label: None,
            required_features: wgpu::Features::empty(), //The device you have limits the features you can use
            required_limits: wgpu::Limits::default(), //The limits field describes the limit of certain types of resource we can create
        },
        None,
    )
    .await
    .unwrap();
```
Unless the trait methods are async, we must use a `blockon` from an async executor. If `winit` internally also uses an async executor, then we may have two different async executors in use, which may be bad.

**Comment by tgross35** [comment]
I have just been using `block_on` for those two functions, but that is a good point. I wish I better understood why these were async at all, I guess maybe this comes from webgpu? Seems like an unlikely point to have performance wins from a nonblocking API.

**Comment by lukexor** [comment]
On Web targets, async with wgpu is extra problematic as `wasm_bindgen_futures::spawn_local` doesn't support returning any results so it's not clear how to get the resolved futures back to the main thread of execution without passing them through channels or static mutexes.

**Comment by hecrj** [comment]
In `iced`, we leverage async channels and `await` to write the event loop using an actual loop.

Why? Because with a loop, the borrow checker can infer a lot more. For instance, we can safely keep user interfaces (which borrow from application state) alive until a specific event happens. This is not possible with a trait-based approach (not even the closure approach) because it would entail self-references. I shared some more details about this approach in this PR: https://github.com/iced-rs/iced/pull/597.

Therefore, the trait-based approach is just a bit of additional unnecessary complexity for `iced`. We will end up rewrapping each event in the trait methods and sending them through a channel anyways. For this reason, we will be using the deprecated API for now, which will save us the unnecessary `match` and rewrapping.

Is there any particular reason the closure approach is being deprecated? It seems quite strange, since the new API is actually relying the deprecated one internally. Why not keep the original alive and offer the new one as a convenient alternative?

**Comment by kchibisov** [comment]
We need return values and sync callbacks. Some APIs mandate return to do e.g. hittest, sync resize, etc. If you do that async you basically blow up and not sync with the windowing systym resulting in visual artifacts, etc. callbacks don't scale at all as well and you can not build extensible systems like you can with extension traits.

The deprecated APIs only here to give some leeway, they'll be gone this week or so from master.

Also, the only difference is that there's no `async` keyword in traits, but all winit traits are _async by matter of operations_ that are planned. Some can callback to you, some will deliver resources, it's just, you can not make them `async`/`await` because you **must** preserve the order.

**Comment by hecrj** [comment]
I see. Return values do seem quite handy.

If the deprecated APIs are going away, I'll switch to the trait-based approach.

`iced`'s event loop isn't really concurrent. The event loop future is polled manually after an event is sent. We only leverage `async` to create a resumable loop, but every event is still processed in order; and a proper request/response protocol can be built through the channels.

**Comment by kchibisov** [comment]
There will be `async` shim for the current trait, because web needs it, since it can not _block_ and in the future there could be an option to write your own backend/loop structures.

The traits are also needed to split winit, which you can not do with closures really.

**Comment by thinnerthinker** [comment]
If anyone is interested in my pattern of integrating 0.30.0 into a cross-compiled wgpu app, check out https://github.com/thinnerthinker/sursface/blob/main/examples/src/hello_triangle/main.rs

**Comment by CXCubeHD** [comment]
Ngl with the new update the borrow checker is fucking me so hard rn

**Comment by mickvangelderen** [comment]
Should `run_app` take `A` instead of `&mut A`? This would allow using a wrapper type without introducing another level of redirection. `ApplicationHandler<T>` is already automatically implemented for `&mut A` when `A: ApplicationHandler<T>`. 

```rust
#[inline]
#[cfg(not(all(web_platform, target_feature = "exception-handling")))]
pub fn run_app<A: ApplicationHandler<T>>(self, app: &mut A) -> Result<(), EventLoopError> {
    self.event_loop.run(|event, event_loop| dispatch_event_for_app(app, event_loop, event))
}
```

**Comment by madsmtm** [comment]
> Should `run_app` take `A` instead of `&mut A`?

Yes, and it was done in https://github.com/rust-windowing/winit/pull/3721

**Comment by casey** [comment]
One slightly annoying thing with the new trait is it makes error handling a bit harder. Using the old interface, I had a single `handle_event` function which could produce an error, which I would handle. Using the new trait-based interface, the event handling functions don't return a Result, so I have to perform error handling in each. (Which means, if I want to use `?`, I have to wrap each in an inner function that can return a `Result`.)

One way to make this more ergonomic would be something like this:

```
pub trait ApplicationHandler<T: 'static = (), E> {
    fn handle_error(&mut self, error: E);
    
    fn new_events(&mut self, _: &ActiveEventLoop, _: StartCause) -> Result<(), E>
    …
}
```

`ApplicationHandler` now has a generic `E` error type, which represents a user error, and trait methods now return `Result<(), E>`.

If a trait method returns an error, then `handle_error` is invoked with that error.

This would allow all user functions to return an error, and allow the user to implement custom error handling in a single place, namely the `handle_error` method.

**Comment by John-Nagle** [comment]
To give a sense of the cost this breaking change imposed on users of the Winit crate, I'm attaching the diffs of the changes that were needed to make Rend3 work with the new winit.

[diffs.txt](https://github.com/user-attachments/files/18553094/diffs.txt)

This is a fight over who owns the event loop. Rend3 has an event loop. Winit has an event loop. Nested event loops are seldom happy-making. This code takes some of the fan-out into different events that Winit does, and packages the events back into a single event type to be fed to the Rend3 event loop, which fans them out again. That's a trait, too, and eventually the application fans out events again.

This breaking change cascades. I was trying to do this without breaking the Rend3 API, but minor changes were forced by the Winit changes. So now my three applications that use Rend3 all need to be fixed. (I seem to be the last user of Rend3, and maintain a fork at rend3-hp.) So those need new branches. and testing.

The documentation of this change presents it as low-impact. It isn't.

**Comment by kchibisov** [comment]
I'm not sure what is new here, winit always owned an event loop, unless you use `pump_events` API.

**Comment by partisani** [comment]
# can't we just return to the old api? are there any implications for using old (pre-0.29) versions of the api?

**Comment by John-Nagle** [comment]
It's been a big job. With the new winit, Winit's event loop wants to allocate the window, rather than having it be passed in from the outside. This forces a breaking change on Rend3's API, breaking all Rend3-using programs. I haven't implemented that part yet.

**Comment by kchibisov** [comment]
>Winit's event loop wants to allocate the window, rather than having it be passed in from the outside

if you create window yourself, there's no point in using winit at this point.

>can't we just return to the old api? are there any implications for using old (pre-0.29) versions of the api?

you can use and maintain whatever you like.

**Comment by John-Nagle** [comment]
> if you create window yourself, there's no point in using winit at this point.

That's the way winit used to work. Currently, that's deprecated, and soon it will go away. It's a breaking change. It's not a bad idea, but it generates work for users.

**Comment by kchibisov** [comment]
Worked when? `Window::new` was always the case, you were not able to create it yourself and pass to winit. The new design would resemble the current use to some extent, in terms of interacting with graphics APIs we have around in rust nothing will really change, only e.g. `set_title` stuff will change a bit.

**Comment by mickvangelderen** [comment]
> This needs better documentation.

Or [even more detailed traits](https://github.com/mickvangelderen/winit-ext/blob/main/examples/winit-ext.rs), although that might be too much.

**Comment by John-Nagle** [comment]
Lots of callbacks vs. events you have to match is kind of a matter of taste. I have to turn some of those callbacks back into events to pass to the outer event loop in Rend3. But I wouldn't change anything. Enough churn already.

How all this works around suspend/resume could use more documentation. Clean suspend/resume is always tough, and cross-platform is even worse. Typical questions: Do you get a "suspended" event when going down? Are you guaranteed that shutdown won't happen before you return from the suspended event? Is suspend/resume always correctly ordered and symmetrical?  Can you postpone suspend a bit if you need to? Does this work the same across all supported platforms?

(Use case: User is logged into an MMO-type game and they close the lid on their laptop. Need to do a clean log out.)

**Comment by mickvangelderen** [comment]
> Lots of callbacks vs. events you have to match is kind of a matter of taste. I have to turn some of those callbacks back into events to pass to the outer event loop in Rend3. But I wouldn't change anything. Enough churn already.

I understood that the motivation for moving from events to callbacks was that some events on some platforms would prefer control to return to the caller only after the event has been processed. I believe this was important for resize events on windows, but I've forgotten the details. It is not obvious how to achieve this through the event based method. It doesn't seem, like you mention, just a matter of taste.

**Comment by Kl4rry** [comment]
Having to create the window after the event loop has started is rather awkward because the window then has to be placed inside an `Option` which is always `Some` except for the first iteration of the eventloop. It just makes it more noisy to access the window in the rest of the application.

**Comment by kchibisov** [comment]
but creating it before event loop is broken and sometimes impossible (android now doesn't even create anything just fyi and macOS doesn't setup methods correctly on a window).

We understand that for simple applications it matters since they must use unwrap, but there probably will be a way around that, it's not like unfixable.

**Comment by RobertBrewitz** [comment]
I've been reiterating my game library code using the trait impl for about a month and a half on the side, it has been a struggle, now in hindsight it's pretty decent though. The biggest struggle was the borrow checker and me having to use self a lot, but that was a "me issue", not a "winit issue"; Still, I do prefer the callback API from a DX perspective 😄

Excerpt:

```rust
pub struct Runtime {
    life_cycle: RuntimeLifeCycle,
    simulation: Simulation,
    input: InputState,
    window_manager: WindowManager,
    pub event_loop_proxy_wrapper: Option<EventLoopProxy<RuntimeEvent>>,
    render_2d: Render2d,
    game: Game,
}
```

Update: I've simplified my use of the ApplicationHandler a bit and eliminated the generics, the most annoying part right now is the inevitable use of `self` a lot and running into borrowing issues because everything has to be attached to the ApplicationHandler to update application state.

```rust
            WindowEvent::RedrawRequested => {
                if let Some(window_ref_mut) = self
                    .window_manager
                    .winit_windows
                    .windows
                    .get_mut(&window_id)
                {
                    window_ref_mut.request_redraw();
                }

                self.input.update();
                self.simulation.update();

                let sim_fixed_steps = &self.simulation.time.sim_fixed_steps.clone();
                let sim_fixed_delta = &self.simulation.time.sim_fixed_delta.clone();
                let sim_frame_delta = &self.simulation.time.sim_frame_delta.clone();
                let render_alpha = &self.simulation.time.render_alpha.clone();

                let mut extracted = Extracted {
                    simulation: &mut self.simulation,
                    input: &mut self.input,
                    window_manager: &mut self.window_manager,
                    event_loop_proxy: self.event_loop_proxy_wrapper.as_ref(),
                };

                self.game.input(&mut extracted);
                self.game
                    .fixed_update(&mut extracted, sim_fixed_steps, sim_fixed_delta);
                self.game.update(&mut extracted, sim_frame_delta);

                if let Some(ctx_ref_mut) = self.render_2d.render_contexts.get_mut(&window_id) {
                    self.game.render(ctx_ref_mut, &extracted, render_alpha);
                }

                self.render_2d.render();
                self.input.end_frame();
            }
```
